src/provider/systemdmanager/manager.rs

```diff
@@ -291,12 +298,44 @@ impl SystemdManager {
     pub async fn stop(&self, unit: &str) -> anyhow::Result<()> {
         debug!("Trying to stop systemd unit [{}]", unit);
 
-        match self.proxy.stop_unit(unit, StopMode::Fail).await {
-            Ok(result) => {
-                debug!("Successfully stopped service [{}]: [{:?}]", unit, result);
-                Ok(())
-            }
-            Err(e) => Err(anyhow!("Error stopping service [{}]: {}", unit, e)),
+        let result = self
+            .process_job(|proxy| proxy.stop_unit(unit, StopMode::Fail))
+            .await;
+
+        if result.is_ok() {
+            debug!("Successfully stopped service [{}]", unit);
+        }
+
+        result.map_err(|e| anyhow!("Error stopping service [{}]: {}", unit, e))
+    }
+
+    /// Runs the given task and waits until the job returned by the task
+    /// is finished.
+    async fn process_job<'a, F, Fut>(&'a self, task: F) -> anyhow::Result<()>
+    where
+        F: Fn(&'a AsyncManagerProxy) -> Fut,
+        Fut: Future<Output = zbus::Result<AsyncJobProxy<'a>>>,
+    {
+        let signals = self
+            .proxy
+            .receive_signal(ManagerSignals::JobRemoved.into())
```

Comment:
I've tried looking at the zbus documentation briefly and I can't find any good documentation about what receive_signal does and the whole thing being generated by macros makes it hard to see as well.

Could you please document what's going on in this method?

Reply:
Done. Please have a look if the comments are sufficient.

Reply:
Thank you and I'm sorry for not making myself clearer.

I'd have liked documentation like this (obviously in "correct"):

`Systemd sends a JobRemoved signal when bla bla, we need to wait for this signal before we can return a success because otherwise XYZ will happen which will cause the issue #136. If this signal never comes .... bla bla"

Basically describing _why_ you are doing it this way and not in any other way.

Reply:
I described now shortly the systemd behaviour.